rfc/004-priority-based-mempool.md

```diff
@@ -0,0 +1,145 @@
+
+In case phase 0 is not enough (for reference, go-ethereum uses [a heap](https://github.com/ethereum/go-ethereum/blob/6c9f040ebeafcc680b0c457e6f4886e2bca32527/core/tx_list.go#L440) to do similar logic), we may want to change the underlying data structure storing transactions. However, before adding a new data structure to replace existing CList implementation, certain refactoring work needs to be finished.
+
+Goal:
+
+1. No functionality change and no performance improvement. This phase should be to design a common underlying mempool interface to allow different implementations
+
+Existing mempool code has many places tightly coupled with CList implementation (like in mempool reactor), and this phase mainly works as a middle ground to abstract away this coupling to allow future mempool implementation using other data structures for better performances.
+
+Proposed interface changes:
+
+```golang
+type basemempool struct {
+    mempoolImpl
+
+    // Atomic integers
+    height   int64 // the last block Update()'d to
+    txsBytes int64 // total size of mempool, in bytes
+
+    // Notify listeners (ie. consensus) when txs are available
+    notifiedTxsAvailable bool
+    txsAvailable         chan struct{} // fires once for each height, when the mempool is not empty
+
+    // Keep a cache of already-seen txs.
+    // This reduces the pressure on the proxyApp.
+    cache     txCache
+    preCheck  PreCheckFunc
+    postCheck PostCheckFunc
+
+    config *cfg.MempoolConfig
+
+    // Exclusive mutex for Update method to prevent concurrent execution of
+    // CheckTx or ReapMaxBytesMaxGas(ReapMaxTxs) methods.
+    updateMtx sync.RWMutex
+
+    wal          *auto.AutoFile // a log of mempool txs
+    proxyAppConn proxy.AppConnMempool
+
+    logger  log.Logger
+    metrics *Metrics
+}
+
+type mempoolImpl interface {
+    Size() int
+
+    addTx(*mempoolTx, uint64)
+    removeTx(types.Tx) bool // return whether corresponding element is removed or not
+    updateRecheckCursor()
+    reapMaxTxs(int) types.Txs
+    reapMaxBytesMaxGas(int64, int64) types.Txs // based on priority
+    recheckTxs(proxy.AppConnMempool)
+    isRecheckCursorNil() bool
+    getRecheckCursorTx() *mempoolTx
+    getMempoolTx(types.Tx) *mempoolTx
+    deleteAll()
+    // ...and more
+}
+```
+
+Such that mempool's shared code will live under `basemempool` (callback handling, WAL, etc.) while different `mempoolImpl` only needs to implement required methods on transaction addition / removal / iteration etc. The proposed interface is [implemented in this code base](https://github.com/QuarkChain/tendermintx/blob/master/mempool/mempool.go).
+
+### Phase 2: Implement mempool based on different data structures
+
+We will have more discussions on what data structure fits the requirements best. For reference, we have some preliminary profiling results on [left-leaning red–black tree and BTree](https://github.com/QuarkChain/tendermintx/blob/master/mempool/bench_test.go#L18-L31).
```

Comment:
Why do we need anything fancy here? Can we not just use a max heap?

Reply:
You want a data structure that can be added to in parallel in a lock-free fashion. Neither max heaps or RB trees satisfy this to my knowledge. 

I am still in favor of the concurrent lock-free AVL variant from: http://ddana.cswp.cs.technion.ac.il/wp-content/uploads/sites/19/2015/12/logicalorderingavl.pdf

ref: https://github.com/tendermint/tendermint/issues/2147#issue-347673061

Reply:
1. using balance tree can make the iteration during reaping mempool O(n) instead of O(nlogn) as it’s just an in-order traversal, while cost would be amortized during insertions. essentially a trade-off between performances of *tx addition* and *reaping*
1. a lock-free data structure supporting parallel addition sounds really interesting. but considering the trade-off between performance gain and implementation complexity, I think some profilings can help get more insights. I would say if a simpler lock-based data structure can provide similar performance as current `CList`  then it’s ok to have a lock

Reply:
> I am strongly in favor of the concurrent lock-free AVL variant from: http://ddana.cswp.cs.technion.ac.il/wp-content/uploads/sites/19/2015/12/logicalorderingavl.pdf

are there any implementations?

Reply:
> a lock-free data structure supporting parallel addition sounds really interesting.

yeah. also, being able to Reap transactions from the mempool and, in the same time, iterate over the tree (broadcasting transactions) would be nice too.

Reply:
Maybe I'm wrong here, but aren't `CheckTx` calls synchronous and serialized? So why do we need a lock-free data structure when the access to it will be serialized by `CheckTx` anyway?

Reply:
> Maybe I'm wrong here, but aren't CheckTx calls synchronous and serialized?

they're supposed to be async https://github.com/tendermint/tendermint/blob/d1a88fe39f0200b89b406d783a5a2490cfe3c42a/mempool/clist_mempool.go#L291

Reply:
> Maybe I'm wrong here, but aren't CheckTx calls synchronous and serialized? So why do we need a lock-free data structure when the access to it will be serialized by CheckTx anyway?

The calls are made asynchronously, but in serial order. But we may want to perform two other operations simultaneously:
- broadcast txs to peers (requires reading from the set of txs)
- reaping txs for a proposal block (requires reading an ordered batch of txs)